kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java

```diff
@@ -0,0 +1,283 @@
+
+  // Configuration for the plugin.
+  private final Config sconfig;
+  
+  // Static constants for configuring Kafka producer. 
+  private static final String BROKER_LIST = "bootstrap.servers";
+  private static final String KEY_SERIALIZER = "key.serializer";
+  private static final String VAL_SERIALIZER = "value.serializer";
+  private static final String CLIENT_ID = "client.id";
+  private static final String ACKS_REQUIRED = "request.required.acks";
+
+  // Kafka properties
+  private final Properties props = new Properties();
+  
+  // Kafka producer configuration
+  private ProducerConfig config;
+  
+  // Kafka producer handle
+  private org.apache.kafka.clients.producer.KafkaProducer<String, String> producer;
+  
+  // Plugin context
+  private RealtimeContext context;
+  
+  // Optimization to collect fields extracted. This is required because the schema
+  // is not available during initialization and configuration phase.
+  private boolean fieldsExtracted = false;
+  
+  // If Async mode
+  private boolean isAsync = false;
+  
+  // List of Kafka topics.
+  private String[] topics;
+  
+  // required for testing.
+  public KafkaProducer(Config config) {
+    this.sconfig = config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+    
+    if(!sconfig.isAsync.equalsIgnoreCase("true") && !sconfig.isAsync.equalsIgnoreCase("false")) {
+      throw new IllegalArgumentException("Async flag has to be either TRUE or FALSE.");
+    }
+    
+    // Validations to be added.
+  }
+
+  @Override
+  public void initialize(RealtimeContext context) throws Exception {
+    super.initialize(context);
+    this.context = context;
+
+    // Extract the topics
+    topics = sconfig.topics.split(",");
+    
+    // Configure the properties for kafka.
+    props.put(BROKER_LIST, sconfig.brokers);
+    props.put(KEY_SERIALIZER, "org.apache.kafka.common.serialization.StringSerializer");
+    props.put(VAL_SERIALIZER, "org.apache.kafka.common.serialization.StringSerializer");
+    props.put(CLIENT_ID, "kafka-producer-" + context.getInstanceId());
+    if (sconfig.isAsync.equalsIgnoreCase("TRUE")) {
+      props.put(ACKS_REQUIRED, "1");
+      isAsync = true;
+    }
+    
+    //config = new ProducerConfig(props);
+    producer = new org.apache.kafka.clients.producer.KafkaProducer<String, String>(props);
+    
+  }
+  
+  @Override
+  public int write(Iterable<StructuredRecord> objects, final DataWriter dataWriter) throws Exception {
+    int count = 0;
+
+    List<Schema.Field> fields = Lists.newArrayList();
+    
+    // For each object
+    for (StructuredRecord object : objects) {
+      
+      // Extract the field names from the object passed in. This is required
+      // because this information is not available in initialize or configuration phase.
+      if(!fieldsExtracted) {
+        fields = object.getSchema().getFields();
```

Comment:
Actually we don't need to memorize the `fields` with `fieldsExtracted`. Getting fields from schema is pretty much free.